content/chapters/io/lab/content/file-descriptors.md

```diff
@@ -0,0 +1,234 @@
+# File Descriptors
+
+You've seen from running the code above that `open()` returns a **number**.
+This number is a **file descriptor**.
+Run the code above multiple times.
+You'll always get file descriptor 3.
+We can already tell these numbers have some meaning and aren't just some seemingly random numbers, like PIDs were.
+
+You've most likely used file descriptors before without knowing.
+Remember the following concepts:
+
+- `stdin` (standard input)
+- `stdout` (standard output)
+- `stderr` (standard error)
+
+When you wanted to run a shell command and ignore its errors, you wrote something like:
+
+```console
+student@os:~$ ls 2> /dev/null
+```
+
+The command above uses `2> /dev/null` to **redirect** `stderr` to /dev/null.
+
+[Quiz](../quiz/stderr-fd.md)
+
+Good, so we know that `stderr` is file descriptor 2.
+The code in `support/file-descriptors/open_directory.c` opened the directory as file descriptor 3.
+So what are file descriptors 0 and 1?
+They are `stdin` and `stdout`, respectively.
+
+Now we know that the basic I/O streams of a process are correspond to specific file decriptors.
+But what is a file descriptor exactly?
+From the application's point of view, it is a positive integer acting as a unique identifier for an I/O channel, such as a file.
+To the operating system, a file descriptor is an index.
+Each process has what's called a **file descriptor table**, which is just a fancy name for an array of pointers.
+A file descriptor is an index in this array.
+Each element in this array points towards a structure in the kernel's memory that represents an I/O resource.
+This structure is called the **open file strcuture**.
+
+To illustrate all this information, look at the image below.
+All data structures shown below are stored in the Kernel's memory area and are transparent from the user space.
+And remember that the file descriptor table is bound to the process, so all its threads share the same file descriptors.
+
+![File Descriptors](../media/file-descriptors.svg).
+
+To find out more about the contents of these structures, check out [this section in the Arena](./arena.md#open-file-structure-in-the-kernel)
```

Comment:
No. The Arena is a collection of more complex /or not super-relevant, but interesting exercises. It can be linked anywhere.

Reply:
The reference itself is not the issue, but the wording may encourage readers to go to the Arena section instead of focusing on the core information. It may create more context switches, without much gain.